src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import React, { useEffect } from 'react';
+import { i18n } from '@osd/i18n';
+import { EuiComboBox } from '@elastic/eui';
+import { DataSourceType, GenericDataSource } from '../datasource_services';
+import { DataSourceGroup, DataSourceSelectableProps } from './types';
+import { ISourceDataSet, IndexPatternOption } from '../datasource';
+
+type DataSourceTypeKey = 'DEFAULT_INDEX_PATTERNS' | 's3glue' | 'spark';
+
+// Mapping between datasource type and its display name.
+const DATASOURCE_TYPE_DISPLAY_NAME_MAP: Record<DataSourceTypeKey, string> = {
+  DEFAULT_INDEX_PATTERNS: 'Index patterns',
+  s3glue: 'Amazon S3',
+  spark: 'Spark',
+};
```

Comment:
Every time I review this piece i get triggered; I feel there is something very wrong happening here. 

1. I see `DEFAULT_INDEX_PATTERNS` defined and exported in `src/plugins/data/public/data_sources/register_default_datasource.ts`. That should be imported here. 
2. I see `registerDataSourceType()` being called only once and that too on `DEFAULT_INDEX_PATTERNS`; where are `s3glue` and `spark` coming from?
3. Looking at the signature of `registerDataSourceType()`, it appears though that the goal is to let the consumers define _ANY_ type that they want. If that is true, why are we limiting them to these 3 here? Also, who ever is registering the type should be responsible for providing a display name for it as well. We cannot maintain a repository of _ANY_ type to a display name on OSD.
4. The display name or `DEFAULT_INDEX_PATTERNS` is provided as `Index patterns` here but `src/plugins/data/public/data_sources/register_default_datasource.ts` in fact reads from `i18n('data.datasource.type.openSearchDefault')` and defaults to `OpenSearch Default`. Why are they not the same?
5. If the goal is for OSD to lock data source types to these 3 and not allow any other type to be defined, why doesn't the factory have any guards enforced? Use `enum`, proper typing, and checks in `registerDataSourceType`.
6. Looking at the code, `s3glue` and `spark` appear to be keywords. If based on the above points, they are here to stay and you put them in an `enum`, please use `S3GLUE` and `SPARK`.

Reply:
re: DEFAULT_INDEX_PATTERNS -- I made a point in discussion a while back...
What makes most sense here is to define a registration for types, and then have OSD register itself the "default" types.  In a way its a sort of reference-implementation for all other plugins.   It also separates the concerns completely, which will lead to cleaner code.

Reply:
i had a fork to make data source registration take in the data source type, where the type contains the display name, backend name, and key. i don't think we are going with that in this PR, but putting here again for reference https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5167#issuecomment-1745873686